streaming/base/dataset.py

```diff
@@ -294,40 +348,79 @@ def __init__(self,
-        self.shard_states_filename = os.path.join(self._filelock_root, '_shard_states_filelock')
+        # Cache usage in bytes.
+        self._cache_usage = SharedScalar(np.int64, f'{self._shm_prefix}_cache_usage')
 
-        # Create or attach shard_states array (tells if each shard is unknown, downloading, or
-        # downloaded).
-        shard_states_shm = CreateSharedMemory(name=f'{self._shm_prefix}_shard_states',
-                                              size=self.num_shards * np.uint8(0).nbytes)
-        self._shard_states = shard_states_shm.shm
+        # Shard states array. Tells if a shard is missing, downloading, or present (eviction
+        # happens under the lock).
+        self._shard_states = SharedArray(self.num_shards, np.uint8,
+                                         f'{self._shm_prefix}_shard_states')
+
+        # Time of last access per shard. This is used to decide which shard(s) to evict when we run
+        # out of space.
+        self._shard_access_times = SharedArray(self.num_shards, np.float64,
+                                               f'{self._shm_prefix}_shard_access_times')
+
+        # Initialize shared memory objects.
+        if world.is_local_leader:
+            # Set initial epoch (before any resumption).
+            self.next_epoch = 0
+
+            # Normalize each stream's local dir, discovering which shards are present.
+            are_shards_present = []
+            for stream in self.streams:
+                stream_shards = stream.get_shards(world)
+                are_shards_present += stream.init_local_dir(stream_shards)
+
+            # Calculate the initial cache usage using shard presence info.
+            #
+            # If we are above cache_limit, do nothing about it until the first download (which will
+            # evict until happy).
+            self.cache_usage = 0
+            for stream in self.streams:
+                self.cache_usage += stream.get_index_size()
+            for shard_id, is_shard_present in enumerate(are_shards_present):
+                if is_shard_present:
+                    stream_id = self.stream_per_shard[shard_id]
+                    stream = self.streams[stream_id]
+                    shard = self.shards[shard_id]
+                    self.cache_usage += shard.get_persistent_size(stream.safe_keep_zip)
+
+            # Also use shard presence to initialize the shard states array and last access times.
+            for shard_id, is_shard_present in enumerate(are_shards_present):
+                if is_shard_present:
+                    self._shard_states[shard_id] = _ShardState.PRESENT
+                    self._shard_access_times[shard_id] = time()
+                else:
+                    self._shard_states[shard_id] = _ShardState.MISSING
+                    self._shard_access_times[shard_id] = NEVER
+
+        self._shared_barrier(world.workers_per_node)
```

Comment:
I see a hang when running a model training on more than one GPU. However, if I replace this line of code with torch dist barrier code then my training works. Looks like  `_shared_barrier` is not sync processes.